### ISSUE TITLE
[java] NoClassDefFoundError on upgrade from 6.25.0 to 6.26.0

### ISSUE THREAD

**Issue by pzygielo** [pmd/pmd#2663]
**Affects PMD Version:** 

[6.26.0](https://github.com/pmd/pmd/releases/tag/pmd_releases%2F6.26.0)

**Description:**

Upgrading pmd-core & pmd-java in maven project causes build to fail with:
```
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-pmd-plugin:3.13.0:pmd (pmd) on project cnfex: Execution pmd of goal org.apache.maven.plugins:maven-pmd-plugin:3.13.0:pmd failed: A required class was missing while executing org.apache.maven.plugins:maven-pmd-plugin:3.13.0:pmd: javax/el/ELResolver
```
PMD 6.25.0 didn't complain about this missing dependency (which isn't required to compile and package given project, with `jakarta.faces-api` in `provided` scope).

**Exception Stacktrace:**
First caused by
```
Caused by: java.lang.NoClassDefFoundError: javax/el/ELResolver
    at java.lang.Class.getDeclaredMethods0 (Native Method)
    at java.lang.Class.privateGetDeclaredMethods (Class.java:2701)
    at java.lang.Class.privateGetPublicMethods (Class.java:2902)
    at java.lang.Class.getMethods (Class.java:1615)
    at net.sourceforge.pmd.lang.java.ast.internal.ImportWrapper.<init> (ImportWrapper.java:44)
    at net.sourceforge.pmd.lang.java.ast.internal.ImportWrapper.<init> (ImportWrapper.java:33)
    at net.sourceforge.pmd.lang.java.rule.bestpractices.UnusedImportsRule.visit (UnusedImportsRule.java:131)
    at net.sourceforge.pmd.lang.java.ast.ASTImportDeclaration.jjtAccept (ASTImportDeclaration.java:143)
    at net.sourceforge.pmd.lang.java.rule.AbstractJavaRule.visit (AbstractJavaRule.java:219)
    at net.sourceforge.pmd.lang.java.rule.AbstractJavaRule.visit (AbstractJavaRule.java:391)
    at net.sourceforge.pmd.lang.java.rule.bestpractices.UnusedImportsRule.visit (UnusedImportsRule.java:57)
    at net.sourceforge.pmd.lang.java.ast.ASTCompilationUnit.jjtAccept (ASTCompilationUnit.java:44)
    at net.sourceforge.pmd.lang.java.rule.AbstractJavaRule.visitAll (AbstractJavaRule.java:162)
    at net.sourceforge.pmd.lang.java.rule.AbstractJavaRule.apply (AbstractJavaRule.java:156)
    at net.sourceforge.pmd.lang.rule.AbstractDelegateRule.apply (AbstractDelegateRule.java:336)
```

**Code Sample demonstrating the issue:**

[pmd-6.26.0-CNFEx](https://github.com/pzrep/pmd-6.26.0-CNFEx/tree/dependabot/maven/pmdVersion-6.26.0)

**Steps to reproduce:**
For the reproducer from above:
```
./mvnw test -DpmdVersion=6.25.0
```
vs
```
./mvnw test -DpmdVersion=6.26.0
```

**Running PMD through:** *Maven*

**Comment by boris-petrov** [comment]
I get something similar but with a different class - `java.lang.NoClassDefFoundError: org/apache/commons/net/ftp/parser/FTPFileEntryParserFactory`.

**Comment by pzygielo** [comment]
> I get something similar but with a different class - `java.lang.NoClassDefFoundError: org/apache/commons/net/ftp/parser/FTPFileEntryParserFactory`.

I also get different classes reported in different projects that have different dependencies. That's why I think it's not Faces or EL to be blamed :grin:

**Comment by pzygielo** [comment]
bisected to 62666213baf4270b5c1567ce3eb09bffe638b09d

**Comment by adangel** [comment]
Thanks for reporting. As always, there are a couple of issues playing together:

* The class `javax.faces.application.Application` from which you inherit, contains a method `getELResolver()` which returns the type `javax.el.ELResolver` - the one which is missing. This means, that the compile classpath is actually incomplete.
* The class `Application` is brought into the compile classpath with the dependency `jakarta.faces:jakarta.faces-api:2.3.2` with scope **provided**. Since it is provided, it's transitive dependencies are ignored (see also https://maven.apache.org/guides/introduction/introduction-to-dependency-mechanism.html#Dependency_Scope : "It is not transitive"). I'm not sure, whether this is correct or not, it is as it is. If you add the following method into your class `C`, you'll see, that we are missing a dependency:
```java
  public void foo() {
      javax.el.ELResolver resolver = this.getELResolver();
      System.out.println(resolver);
  }
```
* The rule UnusedImports uses the changed ImportWrapper class in the commit, you've figured out. The reason, why this missing dependency didn't creep up with 6.25.0 is, that now, we always get all the methods from the type, when we previously did this only for static on demand imports. That's one bug.
* But even if we get the methods of the type, we should catch these exceptions and not fail hard if we can't load a class. That's another bug.
* We retrieve not only static methods/fields, but also non-static methods/fields. I think, that's wrong, and that would be bug 3.

Since there are multiple issues, there are multiple workarounds:
* Declare the following dependency:
```xml
        <dependency>
            <groupId>jakarta.el</groupId>
            <artifactId>jakarta.el-api</artifactId>
            <version>3.0.3</version>
            <scope>provided</scope>
        </dependency>
```
* Don't import `javax.faces.application.Application`. You can do this by specifying the FQCN while extending.
* Stay with PMD 6.25.0

**Comment by pzygielo** [comment]
@adangel Thanks for the details on the case.

> The class `javax.faces.application.Application` from which you inherit, contains a method `getELResolver()` which returns the type `javax.el.ELResolver` - the one which is missing. This means, that the compile classpath is actually incomplete.

I would rather say - it's complete enough. :grin: 

> The class `Application` is brought into the compile classpath with the dependency `jakarta.faces:jakarta.faces-api:2.3.2` with scope **provided**. Since it is provided, it's transitive dependencies are ignored

[Using different (default) scope](https://github.com/pzrep/pmd-6.26.0-CNFEx/compare/higher-scope?expand=1) has no effect on PMD - still CNFEx.

> If you add the following method into your class `C`, you'll see, that we are missing a dependency:
> 
> 
> ```java
>   public void foo() {
>       javax.el.ELResolver resolver = this.getELResolver();
>       System.out.println(resolver);
>   }
> ```

But I'm not using other methods and I trust in container to **provide** all other required classes in runtime.

Thank you for suggested workarounds. 

>   Declare the following dependency:
> ```
>         <dependency>
>             <groupId>jakarta.el</groupId>
>             <artifactId>jakarta.el-api</artifactId>
>             <version>3.0.3</version>
>             <scope>provided</scope>
>         </dependency>
> ```

Yeah... But it would place new classes on classpath available not only for pmd but for code as well.
And this is only for this case of missing el-api, which was used only as example in reproducer. One would have to hunt down bunch of others as well.
:heavy_check_mark: But it is a way to go to add them just for analysis in separate profile/job, not available in regular compilation.

> Don't import `javax.faces.application.Application`. You can do this by specifying the FQCN while extending.

If pmd could only complete its check, or at least point to problematic source file. But as `ELResolver` is not used directly in code - it's hard to locate all cases without knowing the culprit (`javax.faces.application.Application` here). I find it harder than adding new dependency.

**Comment by Fernal73** [comment]
Would it be a good idea to test a pmd release against a list of open source repos to verify that it is ready for release?

I think Checkstyle does something similar.


https://github.com/checkstyle/contribution/tree/master/checkstyle-tester#checkstyle-tester

**Comment by jcamerin** [comment]
I had earlier today opened #2683 and closed it after being pointed to this issue by @pzygielo 

I am experiencing this issue WRT org.apache.commons.fileupload.FileItemFactory being reported as missing.  The class is referenced by CommonsMultipartResolver in spring-web, which does not declare the dependency on commons-fileupload.

I have commons-fileupload declared in my application as a runtime dependency, because it is not needed for compile.

Removing the CommonsMultipartResolver import solved the pmd failure, but obviously isnt feasible. 
Temporary workarounds that worked:
1) Add the commons-fileupload as a pmd dependency
2) Change the scope of commons-fileupload from runtime to compile (This is not desirable due to reasons described by @pzygielo )

**Comment by adangel** [comment]
@jcamerin Yes, that's the same issue.

@jcamerin @pzygielo Are the projects you ran into this issue open-source? Then we could integrate them into our [pmd-regression-tester](https://github.com/pmd/pmd-regression-tester). Although it would not be able to detect this special case, because there we need typeresolution, which is not supported in the regression tester yet (see pmd/pmd-regression-tester#48).

The test case is also not so easy to reproduce, because it is about a provided transitive dependency....

**Comment by jcamerin** [comment]
@adangel Unfortunately, no...  this is my employer's project. However, I was able to find the direct path to the issue. If it would be helpful, I believe I could create a small project with little effort that would reproduce our exact issue.

**Comment by adangel** [comment]
It's not about reproducing the issue, it is about having a project that would trigger such and other cases in the future - a smoke test in other words. As a early warning, that we broke something that used to work before. Of course, we could create a new project with that specific problem, but it's useless for a smoke test, if it is artificial (it would only detect this single case).

**Comment by huxi** [comment]
@adangel I had a similar issue with https://github.com/huxi/sulky if that helps.

**Comment by adangel** [comment]
Thanks for the project suggestions. I've created a separate issue for this -> #2693

**Comment by pzygielo** [comment]
I've sanity-checked just released 6.27. It of course works, and I'm quite happy.
Thank you.
